### PR TITLE
refactor: combine Copy To and Copy From

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

1. Introduce common-datasource crate
2. Combine `Copy To` and `Copy From`

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: src/common/datasource/src/error.rs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    #[snafu(display("Invalid url: {}, error :{}", url, source))]
+    InvalidUrl { url: String, source: ParseError },
+
+    #[snafu(display("Failed to build backend, source: {}", source))]
+    BuildBackend {
+        source: object_store::Error,
+        backtrace: Backtrace,
+    },
+
+    #[snafu(display("Failed to lists object in path: {}, source: {}", path, source))]
```
Comment (MichaelScofield) [inline src/common/datasource/src/error.rs:41]:
```suggestion
    #[snafu(display("Failed to list object in path: {}, source: {}", path, source))]
```

File: src/common/datasource/src/lister.rs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+                    .await
+                    .context(error::ListObjectsSnafu { path: &self.path })?;
+
+                streamer
+                    .try_filter(|f| {
+                        let res = if let Some(regex) = &self.regex {
+                            regex.is_match(f.name())
+                        } else {
+                            true
+                        };
```
Comment (MichaelScofield) [inline src/common/datasource/src/lister.rs:65]:
```suggestion
                        let res = matches!(self.regex.map(|x| x.is_match(f.name())), Some(true));
```

File: src/datanode/src/sql/copy_table_from.rs [hunk 3]
```diff
@@ -15,45 +15,56 @@
...
+            build_backend(&req.location, req.connection).context(error::BuildBackendSnafu)?;
+
+        let (dir, filename) = find_dir_and_filename(&path);
+
+        let regex = if let Some(pattern) = req.pattern {
+            let regex = Regex::new(&pattern).context(error::BuildRegexSnafu)?;
+            Some(regex)
+        } else {
+            None
+        };
```
Comment (MichaelScofield) [inline src/datanode/src/sql/copy_table_from.rs:57]:
```suggestion
        let regex = req.pattern.as_ref().map(|x| Regex::new(x)).transpose().context(error::BuildRegexSnafu)?;
```

**Review by MichaelScofield — APPROVED** [review]

File: src/common/datasource/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+edition.workspace = true
+license.workspace = true
+
+[dependencies]
+common-error = { path = "../error" }
+futures.workspace = true
+object-store = { path = "../../object-store" }
+regex = "1.7.1"
+snafu = { version = "0.7", features = ["backtraces"] }
+url = "2.3.1"
```
Comment (v0y4g3r) [inline src/common/datasource/Cargo.toml:13]: style:
```suggestion
regex = "1.7"
snafu.workspace = true
url = "2.3"
```

File: src/common/datasource/src/object_store.rs [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+            url.host_str().map(|s| s.to_string()),
+            url.path().to_string(),
+        )),
+        Err(ParseError::RelativeUrlWithoutBase) => {
+            Ok((FS_SCHEMA.to_string(), None, url.to_string()))
+        }
+        Err(err) => Err(Error::InvalidUrl {
+            url: url.to_string(),
+            source: err,
+        }),
```
Comment (v0y4g3r) [inline src/common/datasource/src/object_store.rs:45]: nit:
```suggestion
        Err(err) => Err(err).context(InvalidUrlSnafu { url }),
```

File: src/common/datasource/src/object_store.rs [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        S3_SCHEMA => {
+            let host = host.context(error::EmptyHostPathSnafu {
+                url: url.to_string(),
+            })?;
+            Ok(build_s3_backend(&host, "/", connection)?)
+        }
+        FS_SCHEMA => Ok(build_fs_backend("/")?),
+
+        _ => error::UnsupportedBackendProtocolSnafu {
+            protocol: schema.to_string(),
```
Comment (v0y4g3r) [inline src/common/datasource/src/object_store.rs:62]: nit:
```suggestion
            protocol: schema,
```

File: src/datanode/src/error.rs [hunk 7]
```diff
@@ -249,6 +256,12 @@ pub enum Error {
         source: regex::Error,
     },
 
+    #[snafu(display("Failed to list objects, source: {}", source))]
+    ListObjects {
+        backtrace: Backtrace,
```
Comment (v0y4g3r) [inline src/datanode/src/error.rs:261]:
```suggestion
        #[snafu(backtrace)]
```

File: src/datanode/src/error.rs [hunk 8]
```diff
@@ -218,7 +219,13 @@ pub enum Error {
 
     #[snafu(display("Failed to build backend, source: {}", source))]
     BuildBackend {
-        source: object_store::Error,
+        source: DataSourceError,
+        backtrace: Backtrace,
```
Comment (v0y4g3r) [inline src/datanode/src/error.rs:223]:
```suggestion
        #[snafu(backtrace)]
        source: DataSourceError,
```

**Review by v0y4g3r — APPROVED** [review]
Some nitpicking.

**Pull request merged**